pkg/certificate/authorization/service.go

```diff
@@ -0,0 +1,82 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package authorization
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/zeebo/errs"
+	"go.uber.org/zap"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"storj.io/storj/pkg/pb"
+)
+
+// ErrService is the default error class for the authorization service.
+var ErrService = errs.Class("authorization service error")
+
+// Service is the authorization service.
+type Service struct {
+	log *zap.Logger
+	db  *DB
+}
+
+// NewService creates a new authorization service.
+func NewService(log *zap.Logger, db *DB) *Service {
+	return &Service{
+		log: log,
+		db:  db,
+	}
+}
+
+// GetOrCreate will return an authorization for the given user ID
+func (service *Service) GetOrCreate(ctx context.Context, req *pb.AuthorizationRequest) (_ *pb.AuthorizationResponse, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	existingGroup, err := service.db.Get(ctx, req.UserId)
+	if err != nil {
+		msg := "error getting authorizations"
+		err = ErrService.Wrap(err)
+		service.log.Error(msg, zap.Error(err))
+		return nil, status.Error(codes.Internal, msg)
+	}
+
+	if len(existingGroup) > 0 {
+		authorization := existingGroup[0]
+		return &pb.AuthorizationResponse{
+			Token: authorization.Token.String(),
+		}, nil
+	}
+
+	createdGroup, err := service.db.Create(ctx, req.UserId, 1)
+	if err != nil {
+		msg := "error creating authorization"
+		err = ErrService.Wrap(err)
+		service.log.Error(msg, zap.Error(err))
+
+		switch err {
+		case ErrCount, ErrEmptyUserID:
```

Comment:
These errors aren't directly returned they are wrapped.

I think that the solution is `errs.Unwrap(err)` and then compare the error, however, I was going to verify it and I found:

1. There is no test case for when this function returns an error (it's tested through the endpoint)
2. It turns that these 2 errors should never happen because
	* `ErrCount` is set to 1
	* `ErrEmptyUserID` if it's empty it will be already be detected by the `service.db.Get` call executed above.

	Nevertheless, I think that's good anyway to have these checks here.

Reply:
Are you suggesting testing the same error cases in the endpoint and the service?